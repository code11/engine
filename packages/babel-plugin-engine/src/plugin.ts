import * as Babel from "@babel/core";
import { VariableDeclaratorVisitor } from "./visitors";

export type PluginConfig = {
  viewLibrary: string;
};

type pluginType = (babel: typeof Babel, state: PluginConfig) => {};

export const plugin: pluginType = (babel, state) => {
  return {
    name: "babel-plugin-engine",
    visitor: {
      VariableDeclarator: VariableDeclaratorVisitor(babel, state),
    },
  };
};
