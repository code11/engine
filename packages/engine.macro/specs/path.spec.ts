import pluginTester from "babel-plugin-tester";
import prettier from "prettier";
import plugin from "babel-plugin-macros";

const macroFile = "'@c11/engine.macro'";
const config = require("./babelConfig.json");

pluginTester({
  plugin,
  babelOptions: { filename: __filename, ...config },
  formatResult: (result: any) => {
    return prettier.format(result, {
      parser: "babel",
    });
  },
  tests: {
    "should support path": {
      code: `
        import { path, view } from ${macroFile};
        const a = path.foo.bar
        const b = path
        const foo: view = ({
          value = path.bar,
          doo = a.bam,
          boo = observe.foo[arg.value][arg.doo],
          moo = observe[b]
        }) => {}
      `,
      snapshot: true,
    },
  },
});
