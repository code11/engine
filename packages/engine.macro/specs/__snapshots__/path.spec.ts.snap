// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unknown plugin should support path: should support path 1`] = `

import { path, view } from '@c11/engine.macro';
const a = path.foo.bar
const b = path
const foo: view = ({
  value = path.bar,
  doo = a.bam,
  boo = observe.foo[arg.value][arg.doo],
  moo = observe[b]
}) => {}

      ↓ ↓ ↓ ↓ ↓ ↓

"use strict";

var _engine = require("@c11/engine.producer");

var _engineView = require("@c11/engine-view");

var a = _engine.path.foo.bar;
var b = _engine.path;
var foo = (0, _engineView.view)({
  args: {
    type: "STRUCT",
    value: {
      value: {
        type: "VALUE",
        value: {
          type: "CONST",
          value: _engine.path.bar,
        },
      },
      doo: {
        type: "VALUE",
        value: {
          type: "CONST",
          value: a.bam,
        },
      },
      boo: {
        type: "OBSERVE",
        path: [
          {
            type: "CONST",
            value: "foo",
          },
          {
            type: "INTERNAL",
            path: ["value"],
          },
          {
            type: "INTERNAL",
            path: ["doo"],
          },
        ],
      },
      moo: {
        type: "OBSERVE",
        path: [
          {
            type: "CONST",
            value: b,
          },
        ],
      },
    },
    meta: {},
  },
  fn: function fn(_ref) {
    var value = _ref.value,
      doo = _ref.doo,
      boo = _ref.boo,
      moo = _ref.moo;
  },
});


`;
